File: 202__Happy_Number.py
class Solution(object):
    def isHappy(self, n):
        """
        :type n: int
        :rtype: bool
        """
        def getsum(n):
            result = 0
            while n:
                result += (n % 10) ** 2
                n /= 10
            return result

        dic = {}
        while n:
            if n in dic:
                return False
            dic[n] = 1
            n = getsum(n)
            if n == 1:
                return True

File: Java/202__Happy_Number.java
class Solution {
    public boolean isHappy(int n) {
        while (true) {
            int next_n = helper(n);
            if (next_n == n) {
                break;
            }
            if (next_n == 4) {
                return false;
            }
            n = next_n;
        }
        if (n == 1) {
            return true;
        } else {
            return false;
        }
    }

    private int helper(int val) {
        int result = 0;
        while (val != 0) {
            int remain = val % 10;
            result += remain * remain;
            val /= 10;
        }
        return result;
    }
}
